crates/but-workspace/src/lib.rs

```diff
@@ -43,6 +59,196 @@ pub fn stacks(gb_dir: &Path) -> Result<Vec<StackEntry>> {
+    pub description: Option<String>,
+    /// The pull(merge) request associated with the branch, or None if no such entity has not been created.
+    /// This is provided by the UI which is also responsible for creating Pull Requests.
+    pub pr_number: Option<usize>,
+    /// A unique identifier for the GitButler review associated with the branch, if any.
+    pub review_id: Option<String>,
+    /// Archived represents the state when series/branch has been integrated and is below the merge base of the branch.
+    /// This would occur when the branch has been merged at the remote and the workspace has been updated with that change.
+    pub archived: bool,
+}
+
+/// Provides the relevant details of a particular [`gitbutler_stack::Stack`]
+/// The entries are ordered from newest to oldest.
+pub fn stack_branches(stack_id: String, ctx: &CommandContext) -> Result<Vec<Branch>> {
+    let state = state_handle(&ctx.project().gb_dir());
+    let default_target = state
+        .get_default_target()
+        .context("failed to get default target")?;
+    let stack_ctx = &ctx.to_stack_context()?;
+
+    let repo = ctx.gix_repository()?;
+    let cache = repo.commit_graph_if_enabled()?;
+    let mut graph = repo.revision_graph(cache.as_ref());
+    let mut check_commit = IsCommitIntegrated::new(ctx, &default_target, &repo, &mut graph)?;
+
+    let mut stack_branches = vec![];
+    let stack = state.get_stack(Id::from_str(&stack_id)?)?;
+    for internal in stack.branches() {
+        let result = convert(
+            stack_ctx,
+            internal,
+            &stack,
+            &default_target,
+            &mut check_commit,
+            stack_branches.clone(),
+        )?;
+        stack_branches.push(result);
+    }
+    stack_branches.reverse();
+    Ok(stack_branches)
+}
+
+fn convert(
+    ctx: &StackContext<'_>,
+    stack_branch: gitbutler_stack::StackBranch,
+    stack: &Stack,
+    default_target: &Target,
+    check_commit: &mut IsCommitIntegrated<'_, '_, '_>,
+    parent_series: Vec<Branch>,
+) -> Result<Branch> {
+    let branch_commits = stack_branch.commits(ctx, stack)?;
+    let remote = default_target.push_remote_name();
+    let mut patches: Vec<Commit> = vec![];
+    let mut is_integrated = false;
+
+    // Reverse first instead of later, so that we catch the first integrated commit
+    for commit in branch_commits.clone().local_commits.iter().rev() {
+        if !is_integrated {
+            is_integrated = check_commit.is_integrated(commit)?;
+        }
+        let api_commit = Commit {
+            id: commit.id().to_gix(),
+            message: commit.message_bstr().into(),
+            has_conflicts: commit.is_conflicted(),
+            state: CommitState::LocalOnly, // TODO: implement this
+        };
+        patches.push(api_commit);
+    }
+    // There should be no duplicates, but dedup because the UI cant handle duplicates
+    patches.dedup_by(|a, b| a.id == b.id);
```

Comment:
I think this only works when the items to be removed are consecutive, so would need to be sorted first. It's probably better to do a `.retain()` and keep track of all seen ids in a hashset.